presto-native-execution/presto_cpp/main/types/FunctionMetadata.cpp

```diff
@@ -0,0 +1,222 @@
+  jsonBasedUdfFunctionMetadata.docString = functionName;
+  jsonBasedUdfFunctionMetadata.schema = kDefaultSchema;
+  jsonBasedUdfFunctionMetadata.outputType =
+      functionSignature.returnType().toString();
+
+  const std::vector<TypeSignature> argumentTypes =
+      functionSignature.argumentTypes();
+  std::vector<std::string> paramTypes;
+  for (const auto& argumentType : argumentTypes) {
+    paramTypes.emplace_back(argumentType.toString());
+  }
+  jsonBasedUdfFunctionMetadata.paramTypes = paramTypes;
+}
+
+const std::vector<protocol::JsonBasedUdfFunctionMetadata>
+getAggregateFunctionMetadata(
+    const std::string& functionName,
+    const std::vector<AggregateFunctionSignaturePtr>&
+        aggregateFunctionSignatures) {
+  std::vector<protocol::JsonBasedUdfFunctionMetadata>
+      jsonBasedUdfFunctionMetadataList;
+  jsonBasedUdfFunctionMetadataList.reserve(aggregateFunctionSignatures.size());
+  const protocol::FunctionKind functionKind = protocol::FunctionKind::AGGREGATE;
+
+  for (const auto& aggregateFunctionSignature : aggregateFunctionSignatures) {
+    protocol::JsonBasedUdfFunctionMetadata jsonBasedUdfFunctionMetadata;
+    jsonBasedUdfFunctionMetadata.functionKind = functionKind;
+    jsonBasedUdfFunctionMetadata.routineCharacteristics =
+        getRoutineCharacteristics(
+            *aggregateFunctionSignature, functionName, functionKind);
+    jsonBasedUdfFunctionMetadata.aggregateMetadata =
+        std::make_shared<protocol::AggregationFunctionMetadata>(
+            getAggregationFunctionMetadata(*aggregateFunctionSignature));
+
+    updateFunctionMetadata(
+        functionName,
+        *aggregateFunctionSignature,
+        jsonBasedUdfFunctionMetadata);
+    jsonBasedUdfFunctionMetadataList.emplace_back(jsonBasedUdfFunctionMetadata);
+  }
+  return jsonBasedUdfFunctionMetadataList;
+}
+
+const std::vector<protocol::JsonBasedUdfFunctionMetadata>
+getScalarFunctionMetadata(
+    const std::string& functionName,
+    const std::vector<const FunctionSignature*>& functionSignatures) {
+  std::vector<protocol::JsonBasedUdfFunctionMetadata>
+      jsonBasedUdfFunctionMetadataList;
+  jsonBasedUdfFunctionMetadataList.reserve(functionSignatures.size());
+  const protocol::FunctionKind functionKind = protocol::FunctionKind::SCALAR;
+
+  for (const auto& functionSignature : functionSignatures) {
+    protocol::JsonBasedUdfFunctionMetadata jsonBasedUdfFunctionMetadata;
+    jsonBasedUdfFunctionMetadata.functionKind = functionKind;
+    jsonBasedUdfFunctionMetadata.routineCharacteristics =
+        getRoutineCharacteristics(
+            *functionSignature, functionName, functionKind);
+
+    updateFunctionMetadata(
+        functionName, *functionSignature, jsonBasedUdfFunctionMetadata);
+    jsonBasedUdfFunctionMetadataList.emplace_back(jsonBasedUdfFunctionMetadata);
+  }
+  return jsonBasedUdfFunctionMetadataList;
+}
+
+const std::vector<protocol::JsonBasedUdfFunctionMetadata>
+getWindowFunctionMetadata(
```

Comment:
A more appropriate name would be "buildWindowJsonMetadata" since we aren't really retrieving anything here. Would be great to use similar naming for aggregate and scalars as well.